File: app/src/main/java/com/proxy/shadowsocksr/impl/SSRLocal.java
package com.proxy.shadowsocksr.impl;

import android.util.Log;

import com.proxy.shadowsocksr.impl.interfaces.OnNeedProtectTCPListener;
import com.proxy.shadowsocksr.impl.plugin.obfs.AbsObfs;
import com.proxy.shadowsocksr.impl.plugin.obfs.ObfsChooser;
import com.proxy.shadowsocksr.impl.plugin.proto.AbsProtocol;
import com.proxy.shadowsocksr.impl.plugin.proto.ProtocolChooser;

import java.net.InetSocketAddress;
import java.nio.ByteBuffer;
import java.nio.channels.ServerSocketChannel;
import java.nio.channels.SocketChannel;
import java.util.HashMap;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public final class SSRLocal extends Thread
{
    private ServerSocketChannel ssc;
    private final String locIP;
    private final String rmtIP;
    private final String pwd;
    private final String cryptMethod;
    private final String tcpProtocol;
    private final String obfsMethod;
    private final String obfsParam;
    private final int rmtPort;
    private final int locPort;

    private volatile boolean isRunning = true;

    private ExecutorService localThreadPool;
    private ExecutorService remoteThreadPool;

    private OnNeedProtectTCPListener onNeedProtectTCPListener;

    private List<String> aclList;

    private final HashMap<String, Object> shareParam;

    public SSRLocal(String locIP, String rmtIP, int rmtPort, int locPort, String pwd,
            String cryptMethod, String tcpProtocol, String obfsMethod, String obfsParam,
            List<String> aclList)
    {
        this.locIP = locIP;
        this.rmtIP = rmtIP;
        this.rmtPort = rmtPort;
        this.locPort = locPort;
        this.pwd = pwd;
        this.cryptMethod = cryptMethod;
        this.tcpProtocol = tcpProtocol;
        this.obfsMethod = obfsMethod;
        this.obfsParam = obfsParam;
        this.aclList = aclList;
        shareParam = new HashMap<>();
    }

    public void setOnNeedProtectTCPListener(
            OnNeedProtectTCPListener onNeedProtectTCPListener)
    {
        this.onNeedProtectTCPListener = onNeedProtectTCPListener;
    }

    class ChannelAttach
    {
        public ByteBuffer localReadBuf = ByteBuffer.allocate(8224);
        public ByteBuffer remoteReadBuf = ByteBuffer.allocate(8224);
        public TCPEncryptor crypto = new TCPEncryptor(pwd, cryptMethod);
        public AbsObfs obfs = ObfsChooser.getObfs(obfsMethod, rmtIP, rmtPort, 1440, obfsParam);
        public AbsProtocol proto = ProtocolChooser
                .getProtocol(tcpProtocol, rmtIP, rmtPort, 1440, shareParam);
        public SocketChannel localSkt;
        public SocketChannel remoteSkt;
        public volatile boolean isDirect = false;//bypass acl list
    }

    @Override public void run()
    {
        localThreadPool = Executors.newCachedThreadPool();
        remoteThreadPool = Executors.newCachedThreadPool();
        //new ThreadPoolExecutor(1, Integer.MAX_VALUE, 300L, TimeUnit.SECONDS, new SynchronousQueue<Runnable>());

        while (isRunning)//When tcp server crashed, restart it.
        {
            try
            {
                ssc = ServerSocketChannel.open();
                //default is block
                ssc.socket().bind(new InetSocketAddress(locIP, locPort));
                while (isRunning)
                {
                    ChannelAttach attach = new ChannelAttach();
                    attach.localSkt = ssc.accept();
                    localThreadPool.execute(new LocalSocketHandler(attach));
                }
            }
            catch (Exception ignored)
            {
            }
            //
            try
            {
                if (ssc != null)
                {
                    ssc.close();
                }
            }
            catch (Exception ignored)
            {
            }
        }
    }

    class LocalSocketHandler implements Runnable
    {
        private final ChannelAttach attach;

        public LocalSocketHandler(ChannelAttach attach)
        {
            this.attach = attach;
        }

        private void handleData() throws Exception
        {
            //ACL Check
            if (aclList.size() != 0)
            {
                int cnt = attach.localSkt.read(attach.localReadBuf);
                if (cnt < 5)
                {
                    return;
                }
                attach.localReadBuf.flip();
                byte atype = attach.localReadBuf.get();
                short port;
                if (atype == (byte) 0x01)
                {
                    Log.e("EXC", "IPV4");
                    byte[] ip = new byte[4];
                    attach.localReadBuf.get(ip);
                    port = attach.localReadBuf.getShort();
                    // TODO need optimize cidr check speed.
                    if (AddressUtils.checkInCIDRRange(AddressUtils.ipv4BytesToInt(ip), aclList))
                    {
                        Log.e("EXC", "IN");
                        attach.isDirect = true;
                        if (!prepareRemote(attach, AddressUtils.ipv4BytesToIp(ip), port))
                        {
                            return;
                        }
                        attach.localReadBuf.clear();
                    }
                    else
                    {
                        Log.e("EXC", "NOT IN");
                        if (!prepareRemote(attach, rmtIP, rmtPort))
                        {
                            return;
                        }
                        attach.localReadBuf.position(cnt);
                        attach.localReadBuf.limit(attach.localReadBuf.capacity());
                    }

                }
                else if (atype == (byte) 0x04)
                {
                    Log.e("EXC", "IPV6");
                    if (!prepareRemote(attach, rmtIP, rmtPort))
                    {
                        return;
                    }
                    attach.localReadBuf.position(cnt);
                    attach.localReadBuf.limit(attach.localReadBuf.capacity());
                    //not ipv6 list yet, but may be bypass loopback ::1, cidr fc00::/7,
                    //and... how to process ipv6 cidr.
                }
                else
                {
                    Log.e("EXC", "DOMAIN");
                    if (!prepareRemote(attach, rmtIP, rmtPort))
                    {
                        return;
                    }
                    attach.localReadBuf.position(cnt);
                    attach.localReadBuf.limit(attach.localReadBuf.capacity());
                }
            }
            else
            {
                //Global mode.
                if (!prepareRemote(attach, rmtIP, rmtPort))
                {
                    return;
                }
            }
            //
            remoteThreadPool.execute(new RemoteSocketHandler(attach));
            //
            while (isRunning)
            {
                if (!checkSessionAlive(attach))
                {
                    Log.e("EXC", "DEAD");
                    break;
                }
                int rcnt = attach.localSkt.read(attach.localReadBuf);
                if (rcnt < 0)
                {
                    break;
                }
                byte[] recv = new byte[attach.localReadBuf.flip()
                                                          .limit()];//size must be limit, not rcnt.
                attach.localReadBuf.get(recv);

                if (!attach.isDirect)
                {
                    recv = attach.proto.beforeEncrypt(recv);
                    recv = attach.crypto.encrypt(recv);
                    recv = attach.obfs.afterEncrypt(recv);
                }

                attach.remoteSkt.write(ByteBuffer.wrap(recv));
                attach.localReadBuf.clear();
            }
        }

        @Override public void run()
        {
            try
            {
                //default is block
                attach.localSkt.socket().setTcpNoDelay(true);
                attach.localSkt.socket().setReuseAddress(true);
                //
                if (!doAuth(attach))
                {
                    Log.e("EXC", "AUTH FAILED");
                    cleanSession(attach);
                    return;
                }
                Log.e("EXC", "AUTH OK");

                if (!processCMD(attach))
                {
                    Log.e("EXC", "CMD FAILED");
                    cleanSession(attach);
                    return;
                }
                Log.e("EXC", "CMD OK");
                handleData();
            }
            catch (Exception ignored)
            {
            }
            cleanSession(attach);
        }
    }

    private boolean doAuth(final ChannelAttach attach) throws Exception
    {
        attach.localReadBuf.limit(1 + 1 + 255);
        int rcnt = attach.localSkt.read(attach.localReadBuf);
        if (rcnt < 3)
        {
            return false;
        }
        attach.localReadBuf.flip();
        if (attach.localReadBuf.get() != (byte) 0x05)//Socks Version
        {
            return false;
        }

        int methodCnt = attach.localReadBuf.get() & 0xFF;
        int mCnt = attach.localReadBuf.limit() - attach.localReadBuf.position();
        if (mCnt < methodCnt || mCnt > methodCnt)
        {
            return false;
        }

        byte[] resp = new byte[]{0x05, (byte) 0xFF};

        while (methodCnt-- != 0)
        {
            if (attach.localReadBuf.get() == 0x00)//Auth_None
            {
                resp[1] = 0x00;
                break;
            }
        }
        attach.localReadBuf.clear();
        attach.localSkt.write(ByteBuffer.wrap(resp));
        return resp[1] == 0x00;
    }

    private boolean processCMD(final ChannelAttach attach) throws Exception
    {
        attach.localReadBuf.limit(3);//Only Read VER,CMD,RSV
        int rcnt = attach.localSkt.read(attach.localReadBuf);
        if (rcnt < 3)
        {
            return false;
        }

        attach.localReadBuf.flip();
        if (attach.localReadBuf.get() != 0x05)//Socks Version
        {
            return false;
        }

        int cmd = attach.localReadBuf.get();
        if (attach.localReadBuf.get() != 0x00)
        {   //RSV must be 0
            return false;
        }

        switch (cmd)
        {
        case 0x01:
            //Response CMD
            attach.localSkt.write(ByteBuffer.wrap(new byte[]{5, 0, 0, 1, 0, 0, 0, 0, 0, 0}));
            attach.localReadBuf.clear();
            return true;
        case 0x03:
            Log.e("EXC", "UDP ASSOC");
            InetSocketAddress isa =
                    ((InetSocketAddress) attach.localSkt.socket().getLocalSocketAddress());
            byte[] addr = isa.getAddress().getAddress();
            byte[] respb = new byte[4 + addr.length + 2];
            respb[0] = 0x05;
            if (isa.getAddress().getHostAddress().contains(":"))
            {
                respb[3] = 0x04;
            }
            else
            {
                respb[3] = 0x01;
            }
            System.arraycopy(addr, 0, respb, 4, addr.length);
            respb[respb.length - 1] = (byte) (locPort & 0xFF);
            respb[respb.length - 2] = (byte) ((locPort >> 8) & 0xFF);
            attach.localSkt.write(ByteBuffer.wrap(respb));
            return true;
        case 0x02://not support BIND
            attach.localSkt.write(ByteBuffer.wrap(new byte[]{5, 7, 0, 0, 0, 0, 0, 0, 0, 0}));
        default:
            return false;
        }
    }

    private boolean prepareRemote(ChannelAttach attach, String remoteIP, int remotePort)
    throws Exception
    {
        attach.remoteSkt = SocketChannel.open();
        //default is block
        attach.remoteSkt.socket().setReuseAddress(true);
        attach.remoteSkt.socket().setTcpNoDelay(true);
        boolean success = onNeedProtectTCPListener.onNeedProtectTCP(attach.remoteSkt.socket());
        if (!success)
        {
            return false;
        }
        attach.remoteSkt.connect(new InetSocketAddress(remoteIP, remotePort));
        return attach.remoteSkt.isConnected();
    }

    private boolean checkSessionAlive(ChannelAttach attach)
    {
        return attach.localSkt != null &&
               attach.remoteSkt != null;
    }

    private void cleanSession(ChannelAttach attach)
    {
        try
        {
            attach.remoteSkt.close();
            attach.localSkt.close();
        }
        catch (Exception ignored)
        {
        }
        attach.remoteSkt = null;
        attach.localSkt = null;
        attach.obfs = null;
        attach.proto = null;
        attach.crypto = null;
        attach.localReadBuf = null;
        attach.remoteReadBuf = null;
    }

    public void stopSSRLocal()
    {
        isRunning = false;
        try
        {
            ssc.close();
        }
        catch (Exception ignored)
        {
        }
        localThreadPool.shutdown();
        ssc = null;
    }

    class RemoteSocketHandler implements Runnable
    {
        private final ChannelAttach attach;

        public RemoteSocketHandler(ChannelAttach attach)
        {
            this.attach = attach;
        }

        @Override public void run()
        {
            try
            {
                while (isRunning)
                {
                    if (!checkSessionAlive(attach))
                    {
                        Log.e("EXC", "DEAD");
                        break;
                    }
                    int rcnt = attach.remoteSkt.read(attach.remoteReadBuf);
                    if (rcnt < 0)
                    {
                        break;
                    }

                    attach.remoteReadBuf.flip();
                    byte[] recv = new byte[rcnt];
                    attach.remoteReadBuf.get(recv);
                    if (!attach.isDirect)
                    {
                        recv = attach.obfs.beforeDecrypt(recv, false);//TODO
                        recv = attach.crypto.decrypt(recv);
                        recv = attach.proto.afterDecrypt(recv);
                    }

                    attach.localSkt.write(ByteBuffer.wrap(recv));
                    attach.remoteReadBuf.clear();
                }
            }
            catch (Exception ignored)
            {
            }
            cleanSession(attach);
        }
    }
}

File: app/src/main/java/com/proxy/shadowsocksr/impl/SSRLocal.kt
package com.proxy.shadowsocksr.impl

import android.util.Log
import com.proxy.shadowsocksr.impl.crypto.CryptoManager
import com.proxy.shadowsocksr.impl.interfaces.OnNeedProtectTCPListener
import com.proxy.shadowsocksr.impl.plugin.obfs.AbsObfs
import com.proxy.shadowsocksr.impl.plugin.obfs.ObfsChooser
import com.proxy.shadowsocksr.impl.plugin.proto.AbsProtocol
import com.proxy.shadowsocksr.impl.plugin.proto.ProtocolChooser
import java.io.InputStream
import java.io.OutputStream
import java.net.InetSocketAddress
import java.net.ServerSocket
import java.net.Socket
import java.nio.channels.ServerSocketChannel
import java.nio.charset.Charset
import java.util.*
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

class SSRLocal(private val  locIP: String, private val rmtIP: String, private val rmtPort: Int, private val locPort: Int, private val  pwd: String,
        private val cryptMethod: String, private val tcpProtocol: String, private val obfsMethod: String, private val obfsParam: String,
        private var isVPNMode: Boolean, private val aclList: List<String>) : Thread()
{
    private var ssc: ServerSocket? = null

    @Volatile private var isRunning: Boolean = true

    private val localThreadPool: ExecutorService = Executors.newCachedThreadPool()
    private val remoteThreadPool: ExecutorService = Executors.newCachedThreadPool()

    var onNeedProtectTCPListener: OnNeedProtectTCPListener? = null

    private val shareParam: HashMap<String, Any> = hashMapOf()

    inner class ChannelAttach
    {
        var localReadBuf: ByteArray? = ByteArray(8224)
        var remoteReadBuf: ByteArray? = ByteArray(8224)
        var crypto: TCPEncryptor? = TCPEncryptor(pwd, cryptMethod)
        var obfs: AbsObfs? = ObfsChooser.getObfs(obfsMethod, rmtIP, rmtPort, 1440, obfsParam,
                shareParam)
        var proto: AbsProtocol? = ProtocolChooser.getProtocol(tcpProtocol, rmtIP, rmtPort, 1440,
                shareParam)
        var localSkt: Socket? = null
        var remoteSkt: Socket? = null
        var localIS: InputStream? = null
        var localOS: OutputStream? = null
        var remoteIS: InputStream? = null
        var remoteOS: OutputStream? = null
        @Volatile var isDirect = false//bypass acl list

        init
        {
            val cryptMethodInfo = CryptoManager.getCipherInfo(cryptMethod)
            val key = ByteArray(cryptMethodInfo[0])
            ImplUtils.EVP_BytesToKey(pwd.toByteArray(Charset.forName("UTF-8")), key)
            shareParam.put("KEY", key)
            shareParam.put("IV LEN", crypto!!.ivLen)
        }
    }

    override fun run()
    {
        while (isRunning)
        {
            try
            {
                ssc = ServerSocket()
                ssc!!.bind(InetSocketAddress(locIP, locPort))
                while (isRunning)
                {
                    val attach = ChannelAttach()
                    attach.localSkt = ssc!!.accept()
                    localThreadPool.execute(LocalSocketHandler(attach))
                }
            }
            catch (ignored: Exception)
            {
            }
            //
            try
            {
                if (ssc != null)
                {
                    ssc!!.close()
                }
            }
            catch (ignored: Exception)
            {
            }
        }
    }

    //
    inner class LocalSocketHandler(val attach: ChannelAttach) : Runnable
    {
        @Throws(Exception::class)
        private fun handleData()
        {
            var recv: ByteArray
            //ACL Check
            if (aclList.size != 0)
            {
                var rcnt = attach.localSkt!!.inputStream.read(attach.localReadBuf, 0,
                        attach.localReadBuf!!.size)
                if (rcnt < 5)
                {
                    return
                }

                val atype = attach.localReadBuf!![0].toInt() and 0xFF
                Log.e("EXC", "ATYPE: " + atype)
                if (atype == 0x01)
                {
                    val ip = Arrays.copyOfRange(attach.localReadBuf, 1, 5)
                    val port: Int = ((attach.localReadBuf!![5].toInt() shl 8) and 0xFF00) or
                            (attach.localReadBuf!![6].toInt() and 0xFF)
                    Log.e("EXC", "PORT: " + port)
                    //
                    // TODO need optimize cidr check speed.
                    if (AddressUtils.checkInCIDRRange(
                            AddressUtils.ipv4BytesToInt(ip), aclList))
                    {
                        attach.isDirect = true
                        if (!prepareRemote(attach, AddressUtils.ipv4BytesToIp(ip), port))
                        {
                            return
                        }
                    }
                    else
                    {
                        if (!prepareRemote(attach, rmtIP, rmtPort))
                        {
                            return
                        }
                    }
                }
                //Currently domain operate and ipv6 operate are the same
                //else if (atype == 0x03)
                //{
                //    if (!prepareRemote(attach, rmtIP, rmtPort))
                //    {
                //        return
                //    }
                //}
                else//0x04
                {
                    if (!prepareRemote(attach, rmtIP, rmtPort))
                    {
                        return
                    }
                    //not ipv6 list yet, but may be bypass loopback ::1, cidr fc00::/7,
                    //and... how to process ipv6 cidr.
                }
                //
                //
                Log.e("EXC", "REMOTE OK")
                remoteThreadPool.execute(RemoteSocketHandler(attach))
                //
                recv = Arrays.copyOfRange(attach.localReadBuf, 0, rcnt)

                if (!attach.isDirect)
                {
                    recv = attach.proto!!.beforeEncrypt(recv)
                    recv = attach.crypto!!.encrypt(recv)
                    recv = attach.obfs!!.afterEncrypt(recv)
                }
                attach.remoteSkt!!.outputStream.write(recv)
            }
            else
            {
                //Global mode.
                if (!prepareRemote(attach, rmtIP, rmtPort))
                {
                    return
                }
                remoteThreadPool.execute(RemoteSocketHandler(attach))
            }
            //
            while (isRunning)
            {
                if (!checkSessionAlive(attach))
                {
                    Log.e("EXC", "DEAD")
                    break
                }
                var rcnt = attach.localSkt!!.inputStream.read(attach.localReadBuf, 0,
                        attach.localReadBuf!!.size)

                if (rcnt < 0)
                {
                    break
                }
                //
                recv = Arrays.copyOfRange(attach.localReadBuf, 0, rcnt)
                ImplUtils.bytesHexDmp("LRECV", recv)
                if (!attach.isDirect)
                {
                    recv = attach.proto!!.beforeEncrypt(recv)
                    recv = attach.crypto!!.encrypt(recv)
                    recv = attach.obfs!!.afterEncrypt(recv)
                }
                attach.remoteSkt!!.outputStream.write(recv)
            }
        }

        override fun run()
        {
            try
            {
                //default is block
                attach.localSkt!!.tcpNoDelay = true
                attach.localSkt!!.reuseAddress = true
                attach.localSkt!!.soTimeout = 600 * 1000
                //attach.localIS = attach.localSkt!!.inputStream
                //attach.localOS = attach.localSkt!!.outputStream
                //
                if (!doAuth(attach))
                {
                    Log.e("EXC", "AUTH FAILED")
                    cleanSession(attach)
                    return
                }
                if (!processCMD(attach))
                {
                    Log.e("EXC", "CMD FAILED")
                    cleanSession(attach)
                    return
                }
                handleData()
            }
            catch (e: Exception)
            {
                e.stackTrace.forEach {
                    Log.e("EXC - lsh","${it.lineNumber} - ${it.toString()}")
                }
            }
            cleanSession(attach)
        }
    }

    @Throws(Exception::class)
    private fun doAuth(attach: ChannelAttach): Boolean
    {
        /* If transplanted to other platforms, you may need to modify this method */
        val rcnt = attach.localSkt!!.inputStream.read(attach.localReadBuf, 0,
                attach.localReadBuf!!.size)
        if (rcnt < 3)
        {
            return false
        }

        val resp = byteArrayOf(0x05, 0x0)
        if (attach.localReadBuf!![0] != 0x05.toByte() ||
                attach.localReadBuf!![1] != 0x01.toByte() ||
                attach.localReadBuf!![2] != 0x00.toByte())
        {
            resp[1] = 0xFF.toByte()
        }

        attach.localSkt!!.outputStream.write(resp, 0, resp.size)
        return resp[1] == 0x0.toByte()
    }

    @Throws(Exception::class)
    private fun processCMD(attach: ChannelAttach): Boolean
    {
        //Only Read VER,CMD,RSV
        val rcnt = attach.localSkt!!.inputStream.read(attach.localReadBuf, 0, 3)
        if (rcnt < 3)
        {
            return false
        }

        if (attach.localReadBuf!![0] != (0x05).toByte() || //Socks Version
                attach.localReadBuf!![2] != (0x00).toByte())//RSV must be 0
        {
            return false
        }

        when (attach.localReadBuf!![1].toInt() and 0xFF)
        {
        //Response CMD
            0x01 ->
            {
                attach.localSkt!!.outputStream.write(byteArrayOf(5, 0, 0, 1, 0, 0, 0, 0, 0, 0), 0,
                        10)
                return true
            }

            0x03 ->
            {
                Log.e("EXC", "UDP ASSOC")
                val isa = (attach.localSkt!!.localSocketAddress) as InetSocketAddress
                val addr = isa.address.address
                val respb = ByteArray(4 + addr.size + 2)
                respb[0] = 0x05
                if (isa.address.hostAddress.contains(":"))
                {
                    respb[3] = 0x04
                }
                else
                {
                    respb[3] = 0x01
                }
                System.arraycopy(addr, 0, respb, 4, addr.size)
                respb[respb.size - 1] = (locPort and 0xFF).toByte()
                respb[respb.size - 2] = ((locPort shr 8) and 0xFF).toByte()
                attach.localSkt!!.outputStream.write(respb)
                return true
            }
        //0x02
            else -> //not support BIND or etc. CMD
            {
                attach.localSkt!!.outputStream.write(byteArrayOf(5, 7, 0, 0, 0, 0, 0, 0, 0, 0))
                return false
            }
        }
    }

    @Throws(Exception::class)
    private fun prepareRemote(attach: ChannelAttach, remoteIP: String, remotePort: Int): Boolean
    {
        attach.remoteSkt = Socket()
        attach.remoteSkt!!.bind(InetSocketAddress(0))
        attach.remoteSkt!!.reuseAddress = true
        attach.remoteSkt!!.tcpNoDelay = true
        attach.remoteSkt!!.soTimeout = 600 * 1000
        //attach.remoteIS = attach.remoteSkt!!.inputStream
        //attach.remoteOS = attach.remoteSkt!!.outputStream
        if (isVPNMode)
        {
            var success = onNeedProtectTCPListener!!.onNeedProtectTCP(attach.remoteSkt!!)
            if (!success)
            {
                return false
            }
        }
        attach.remoteSkt!!.connect(InetSocketAddress(remoteIP, remotePort))
        return attach.remoteSkt!!.isConnected
    }

    private fun checkSessionAlive(attach: ChannelAttach): Boolean =
            attach.localSkt != null && attach.remoteSkt != null

    private fun cleanSession(attach: ChannelAttach)
    {
        try
        {
            attach.remoteSkt!!.close()
        }
        catch (ignored: Exception)
        {
        }
        try
        {
            attach.localSkt!!.close()
        }
        catch(ignored: Exception)
        {
        }
        attach.remoteSkt = null
        attach.localSkt = null
        attach.obfs = null
        attach.proto = null
        attach.crypto = null
        attach.localReadBuf = null
        attach.remoteReadBuf = null
    }

    public fun stopSSRLocal()
    {
        isRunning = false
        try
        {
            ssc!!.close()
        }
        catch (ignored: Exception)
        {
        }
        localThreadPool.shutdown()
        ssc = null
    }

    inner class RemoteSocketHandler(val attach: ChannelAttach) : Runnable
    {
        public override fun run()
        {
            try
            {
                while (isRunning)
                {
                    Log.e("EXC", "RMT")
                    if (!checkSessionAlive(attach))
                    {
                        Log.e("EXC", "DEAD")
                        break
                    }
                    val rcnt = attach.remoteSkt!!.inputStream.read(attach.remoteReadBuf, 0,
                            attach.remoteReadBuf!!.size)
                    if (rcnt < 0)
                    {
                        break
                    }

                    var recv = Arrays.copyOfRange(attach.remoteReadBuf, 0, rcnt)//TODO
                    if (!attach.isDirect)
                    {
                        recv = attach.obfs!!.beforeDecrypt(recv, false)//TODO
                        recv = attach.crypto!!.decrypt(recv)
                        recv = attach.proto!!.afterDecrypt(recv)
                    }

                    ImplUtils.bytesHexDmp("RRECV", recv)

                    attach.localSkt!!.outputStream.write(recv)
                }
            }
            catch (e: Exception)
            {
                e.stackTrace.forEach {
                    Log.e("EXC - rmt","${it.lineNumber} - ${it.toString()}")
                }
            }
            cleanSession(attach)
        }
    }
}
